File: vitalcore-expo/automation/utils/htmlReporter.ts
import * as fs from 'fs';
import * as path from 'path';
import { TestCaseResult } from '../data/testData';
import { Logger } from './logger';

export class HtmlReporter {
  private outputDir: string;

  constructor() {
    this.outputDir = path.join(process.cwd(), 'Test Results', 'HTML');
    if (!fs.existsSync(this.outputDir)) {
      fs.mkdirSync(this.outputDir, { recursive: true });
    }
  }

  public async generateAllReports(results: TestCaseResult[]): Promise<void> {
    Logger.info('Generating HTML Reports...');

    const total = results.length;
    const passed = results.filter((r) => r.status === 'PASS').length;
    const failed = results.filter((r) => r.status === 'FAIL').length;
    const skipped = results.filter((r) => r.status === 'SKIPPED').length;
    const passRate = ((passed / total) * 100).toFixed(1);
    const durationTotal = results.reduce((acc, r) => acc + r.executionTime, 0).toFixed(1);

    const htmlContent = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>VitalCore Appium E2E Execution Report</title>
  <link href="https://fonts.googleapis.com/css2?family=Inter:wght@300;400;600;700;800&display=swap" rel="stylesheet">
  <style>
    :root {
      --bg: #0f172a;
      --card-bg: #1e293b;
      --border: #334155;
      --text: #f8fafc;
      --muted: #94a3b8;
      --primary: #3b82f6;
      --success: #10b981;
      --danger: #ef4444;
      --warning: #f59e0b;
    }
    * { box-sizing: border-box; margin: 0; padding: 0; font-family: 'Inter', sans-serif; }
    body { background: var(--bg); color: var(--text); padding: 2rem; }
    .header { display: flex; justify-content: space-between; align-items: center; margin-bottom: 2rem; border-bottom: 1px solid var(--border); padding-bottom: 1.5rem; }
    .brand { font-size: 1.8rem; font-weight: 800; color: var(--primary); letter-spacing: -0.5px; }
    .subtitle { color: var(--muted); font-size: 0.9rem; margin-top: 0.3rem; }
    .badge { background: rgba(59, 130, 246, 0.15); color: var(--primary); border: 1px solid var(--primary); padding: 0.4rem 0.8rem; borderRadius: 20px; font-size: 0.85rem; font-weight: 600; }
    
    .metrics-grid { display: grid; grid-template-columns: repeat(auto-fit, minmax(200px, 1fr)); gap: 1.25rem; margin-bottom: 2rem; }
    .metric-card { background: var(--card-bg); border: 1px solid var(--border); border-radius: 16px; padding: 1.5rem; }
    .metric-title { font-size: 0.85rem; color: var(--muted); text-transform: uppercase; font-weight: 600; }
    .metric-value { font-size: 2.2rem; font-weight: 800; margin-top: 0.5rem; }
    .text-success { color: var(--success); }
    .text-danger { color: var(--danger); }
    .text-warning { color: var(--warning); }
    .text-primary { color: var(--primary); }

    .table-card { background: var(--card-bg); border: 1px solid var(--border); border-radius: 16px; padding: 1.5rem; overflow-x: auto; }
    .table-header { display: flex; justify-content: space-between; align-items: center; margin-bottom: 1rem; }
    table { width: 100%; border-collapse: collapse; text-align: left; }
    th { padding: 0.75rem 1rem; color: var(--muted); font-size: 0.85rem; border-bottom: 1px solid var(--border); text-transform: uppercase; }
    td { padding: 0.85rem 1rem; border-bottom: 1px solid var(--border); font-size: 0.9rem; }
    tr:hover { background: rgba(255, 255, 255, 0.02); }
    
    .status-tag { padding: 0.25rem 0.6rem; border-radius: 6px; font-weight: 700; font-size: 0.75rem; display: inline-block; }
    .tag-pass { background: rgba(16, 185, 129, 0.15); color: var(--success); }
    .tag-fail { background: rgba(239, 68, 68, 0.15); color: var(--danger); }
    .tag-skip { background: rgba(245, 158, 11, 0.15); color: var(--warning); }

    .filter-bar { display: flex; gap: 0.5rem; }
    .filter-btn { background: var(--bg); border: 1px solid var(--border); color: var(--text); padding: 0.4rem 0.8rem; border-radius: 8px; cursor: pointer; font-size: 0.85rem; }
    .filter-btn.active { background: var(--primary); border-color: var(--primary); }
  </style>
</head>
<body>
  <div class="header">
    <div>
      <div class="brand">VitalCore AI Mobile Automation</div>
      <div class="subtitle">Appium E2E Enterprise Test Execution Dashboard & Analytics</div>
    </div>
    <div class="badge">Environment: Android Emulator (API 33)</div>
  </div>

  <div class="metrics-grid">
    <div class="metric-card">
      <div class="metric-title">Total Test Cases</div>
      <div class="metric-value text-primary">${total}</div>
    </div>
    <div class="metric-card">
      <div class="metric-title">Passed Tests</div>
      <div class="metric-value text-success">${passed}</div>
    </div>
    <div class="metric-card">
      <div class="metric-title">Failed Tests</div>
      <div class="metric-value text-danger">${failed}</div>
    </div>
    <div class="metric-card">
      <div class="metric-title">Skipped Tests</div>
      <div class="metric-value text-warning">${skipped}</div>
    </div>
    <div class="metric-card">
      <div class="metric-title">Pass Rate</div>
      <div class="metric-value text-success">${passRate}%</div>
    </div>
    <div class="metric-card">
      <div class="metric-title">Total Duration</div>
      <div class="metric-value text-primary">${durationTotal}s</div>
    </div>
  </div>

  <div class="table-card">
    <div class="table-header">
      <h2>Executable Test Suite Results (${total} Scenarios)</h2>
      <div class="filter-bar">
        <button class="filter-btn active">All (${total})</button>
        <button class="filter-btn">Passed (${passed})</button>
        <button class="filter-btn">Failed (${failed})</button>
      </div>
    </div>
    <table>
      <thead>
        <tr>
          <th>Test ID</th>
          <th>Module</th>
          <th>Test Name</th>
          <th>Priority</th>
          <th>Status</th>
          <th>Duration</th>
          <th>Actual Result / Details</th>
        </tr>
      </thead>
      <tbody>
        ${results
          .map(
            (r) => `
        <tr>
          <td style="font-family: monospace; font-weight: bold;">${r.id}</td>
          <td>${r.module}</td>
          <td>${r.name}</td>
          <td><span class="status-tag" style="background: #334155;">${r.priority}</span></td>
          <td><span class="status-tag ${
            r.status === 'PASS'
              ? 'tag-pass'
              : r.status === 'FAIL'
              ? 'tag-fail'
              : 'tag-skip'
          }">${r.status}</span></td>
          <td>${r.executionTime}s</td>
          <td style="color: ${r.status === 'FAIL' ? '#f87171' : 'inherit'};">${r.failureReason || r.actualResult}</td>
        </tr>`
          )
          .join('')}
      </tbody>
    </table>
  </div>
</body>
</html>`;

    fs.writeFileSync(path.join(this.outputDir, 'execution-report.html'), htmlContent);
    fs.writeFileSync(path.join(this.outputDir, 'dashboard.html'), htmlContent);
    fs.writeFileSync(path.join(this.outputDir, 'trends.html'), htmlContent);

    Logger.info('HTML Reports generated: execution-report.html, dashboard.html, trends.html');
  }
}

File: testing/selenium/utils/html_reporter.py
from datetime import datetime
from .config import HTML_REPORT_PATH

def generate_html_report(test_results, total_duration_seconds):
    """
    Generates a clean HTML report summarizing test execution results.
    """
    total_tests = len(test_results)
    passed_tests = sum(1 for item in test_results if item['status'] == 'PASS')
    failed_tests = sum(1 for item in test_results if item['status'] == 'FAIL')
    skipped_tests = sum(1 for item in test_results if item['status'] == 'SKIPPED')
    
    pass_pct = round((passed_tests / total_tests * 100), 2) if total_tests > 0 else 0.0

    rows_html = ""
    for item in test_results:
        st = item.get("status", "PASS")
        status_bg = "#dcfce7" if st == "PASS" else ("#fee2e2" if st == "FAIL" else "#fef3c7")
        status_color = "#166534" if st == "PASS" else ("#991b1b" if st == "FAIL" else "#92400e")
        
        err_msg = item.get("error", "")
        err_html = f"<div class='error-msg'>{err_msg}</div>" if err_msg else "-"
        
        screenshot = item.get("screenshot", "")
        img_html = f"<a href='{screenshot}' target='_blank'>View Screenshot</a>" if screenshot else "-"

        rows_html += f"""
        <tr>
            <td><code>{item.get('test_id', '-')}</code></td>
            <td>{item.get('module', '-')}</td>
            <td><strong>{item.get('name', '-')}</strong><br><small>{item.get('description', '')}</small></td>
            <td><span class='badge' style='background:{status_bg}; color:{status_color}'>{st}</span></td>
            <td>{round(item.get('duration', 0), 2)}s</td>
            <td>{img_html}</td>
            <td>{err_html}</td>
        </tr>
        """

    html_content = f"""<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>VitalCore Selenium Test Report</title>
    <style>
        body {{ font-family: 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif; background-color: #f8fafc; color: #0f172a; margin: 0; padding: 24px; }}
        .container {{ max-width: 1200px; margin: 0 auto; background: #ffffff; padding: 32px; border-radius: 16px; box-shadow: 0 4px 6px -1px rgba(0,0,0,0.1); }}
        h1 {{ margin-top: 0; color: #1e293b; font-size: 24px; font-weight: 700; }}
        .metrics-grid {{ display: grid; grid-template-columns: repeat(auto-fit, minmax(180px, 1fr)); gap: 16px; margin: 24px 0; }}
        .metric-card {{ background: #f1f5f9; padding: 16px; border-radius: 12px; text-align: center; }}
        .metric-value {{ font-size: 28px; font-weight: 800; margin-top: 4px; }}
        .metric-label {{ font-size: 12px; color: #64748b; font-weight: 600; text-transform: uppercase; }}
        table {{ width: 100%; border-collapse: collapse; margin-top: 24px; font-size: 14px; }}
        th, td {{ padding: 12px 16px; text-align: left; border-bottom: 1px solid #e2e8f0; vertical-align: top; }}
        th {{ background-color: #0f172a; color: #ffffff; font-weight: 600; }}
        .badge {{ padding: 4px 10px; border-radius: 9999px; font-size: 12px; font-weight: 700; display: inline-block; }}
        .error-msg {{ color: #dc2626; font-size: 12px; max-width: 300px; word-break: break-word; font-family: monospace; }}
    </style>
</head>
<body>
    <div class="container">
        <h1>🧪 VitalCore Selenium Automation Report</h1>
        <p>Execution Date: <strong>{datetime.now().strftime('%B %d, %Y - %H:%M:%S')}</strong> | Duration: <strong>{round(total_duration_seconds, 2)}s</strong></p>

        <div class="metrics-grid">
            <div class="metric-card"><div class="metric-label">Total Tests</div><div class="metric-value">{total_tests}</div></div>
            <div class="metric-card" style="background:#dcfce7;"><div class="metric-label" style="color:#166534">Passed</div><div class="metric-value" style="color:#166534">{passed_tests}</div></div>
            <div class="metric-card" style="background:#fee2e2;"><div class="metric-label" style="color:#991b1b">Failed</div><div class="metric-value" style="color:#991b1b">{failed_tests}</div></div>
            <div class="metric-card" style="background:#fef3c7;"><div class="metric-label" style="color:#92400e">Skipped</div><div class="metric-value" style="color:#92400e">{skipped_tests}</div></div>
            <div class="metric-card"><div class="metric-label">Pass Rate</div><div class="metric-value" style="color:#2563eb">{pass_pct}%</div></div>
        </div>

        <table>
            <thead>
                <tr>
                    <th>Test ID</th>
                    <th>Module</th>
                    <th>Test Case</th>
                    <th>Status</th>
                    <th>Time</th>
                    <th>Screenshot</th>
                    <th>Error</th>
                </tr>
            </thead>
            <tbody>
                {rows_html}
            </tbody>
        </table>
    </div>
</body>
</html>
"""
    with open(HTML_REPORT_PATH, "w", encoding="utf-8") as f:
        f.write(html_content)
    return HTML_REPORT_PATH
